### ISSUE TITLE
SRUN memory error on nosetests

### ISSUE THREAD

**Issue by altaetran** [deepchem/deepchem#228]
Hi,

I just ran the standard nosetests on a gpu node on my cluster, and obtained a memory failure. 

> Test of singletask CoulombMatrixEig regression on .sdf file. ... /home/stanford/pande/users/hraut/src/deepchem/deepchem/datasets/**init**.py:885: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
>   if y[ind, task] == "":
> ok
> Straightforward test of Keras multitask deepchem classification API. ... I tensorflow/core/common_runtime/gpu/gpu_init.cc:102] Found device 0 with properties:
> name: Tesla K80
> major: 3 minor: 7 memoryClockRate (GHz) 0.8235
> pciBusID 0000:90:00.0
> Total memory: 11.25GiB
> Free memory: 11.13GiB
> I tensorflow/core/common_runtime/gpu/gpu_init.cc:126] DMA: 0
> I tensorflow/core/common_runtime/gpu/gpu_init.cc:136] 0:   Y
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> I tensorflow/core/common_runtime/gpu/pool_allocator.cc:244] PoolAllocator: After 3442 get requests, put_count=3284 evicted_count=1000 eviction_rate=0.304507 and unsatisfied allocation rate=0.365485
> I tensorflow/core/common_runtime/gpu/pool_allocator.cc:256] Raising pool_size_limit_ from 100 to 110
> ERROR
> Straightforward test of Tensorflow multitask deepchem classification API. ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> /global/software/extras/tensorflow/0.9/numpy/lib/function_base.py:2487: RuntimeWarning: Degrees of freedom <= 0 for slice
>   warnings.warn("Degrees of freedom <= 0 for slice", RuntimeWarning)
> ok
> Test of hyperparam_opt with singletask RF ECFP regression API. ... /home/stanford/pande/users/hraut/src/deepchem/deepchem/datasets/**init**.py:885: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
>   if y[ind, task] == "":
> ok
> Test of hyperparam_opt with singletask_to_multitask. ... ok
> test_kappa_score (deepchem.metrics.tests.metrics_test.MetricsTest) ... ok
> Test that R^2 metric passes basic sanity tests ... ok
> testAddBias (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testBatchNormalization (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testBatchNormalizationInference (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testBatchNormalizationWithConv (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testBatchNormalizationWithMask (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testBatchNormalizationWithMaskAndConv (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testFullyConnectedLayer (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testMultitaskLogits (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testSoftmaxN (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testSoftmaxNWithNumpy (deepchem.models.tensorflow_models.test.model_ops_test.ModelOpsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> Returns a TensorFlow Session for use in executing tests. ... ok
> testKurtosis (deepchem.models.tensorflow_models.test.utils_test.UtilsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testMean (deepchem.models.tensorflow_models.test.utils_test.UtilsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testMeanWithMask (deepchem.models.tensorflow_models.test.utils_test.UtilsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testMoment (deepchem.models.tensorflow_models.test.utils_test.UtilsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testParseCheckpoint (deepchem.models.tensorflow_models.test.utils_test.UtilsTest) ... ok
> testSkewness (deepchem.models.tensorflow_models.test.utils_test.UtilsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testVariance (deepchem.models.tensorflow_models.test.utils_test.UtilsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> testVarianceWithMask (deepchem.models.tensorflow_models.test.utils_test.UtilsTest) ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ok
> Returns a TensorFlow Session for use in executing tests. ... ok
> Straightforward test of Keras multitask deepchem classification API. ... I tensorflow/core/common_runtime/gpu/gpu_device.cc:806] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:90:00.0)
> ERROR
> Test of singletask RF ECFP regression API. ... /home/stanford/pande/users/hraut/src/deepchem/deepchem/datasets/**init**.py:885: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
>   if y[ind, task] == "":
> ok
> Test of singletask RF ECFP regression API: sharded edition. ... srun: Exceeded job memory limit
> srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
> srun: error: xs-0026: task 0: Killed

Is there any way to avoid having the tests use so much memory? I feel like the tests should not be using more than 12gb of memory. Also, a few of the tests seem to be producing errors, even with the new Keras version.

**Comment by rbharath** [comment]
I'm seeing the memory issues too. Will look into it, but I think it has something to do with the fact that all the `keras` examples run on `tensorflow`. In general, I've noticed that `tensorflow` is quite bad at cleanup. Objects don't release their memory promptly, causing many errors like this.

I'm quite surprised you're seeing keras errors still. Can you try running failing tests in isolation to see if they still fail?

**Comment by rbharath** [comment]
Closing this issue since I think it's no longer a problem.
